src/Services/Dominion/Actions/Military/TrainActionService.php

```diff
@@ -187,4 +168,52 @@ public function train(Dominion $dominion, array $data): array
             ],
         ];
     }
+    
+    /**
+     * Returns training message for a train action.
+     *
+     * @param Dominion $dominion
+     * @param array $unitsToTrain
+     * @param array $totalCosts
+     * @return string
+     */
+    private function getTrainingMessage(Dominion $dominion, $totalCosts) {
+        $unitsToTrainStringParts = [];
+
+       foreach ($unitsToTrain as $unitType => $amount) {
+            if($amount > 0) {
+                $race = strtolower($this->unitHelper->getUnitName($unitType, $dominion->race));
+                $unitsToTrainStringParts[] = number_format($amount) . ' ' . str_plural($race, $amount);
+            }
+       }
+        
+       $unitsToTrainString = $this->generateSentenceFromArray($unitsToTrainStringParts);
+
+       $trainingCostsStringParts = [];
+       foreach($totalCosts as $costType => $cost) {
+           if($cost > 0){
+             $trainingCostsStringParts[] = number_format($cost) . ' ' . str_plural($costType, $cost);
+           }
+       }
+                               
+       $trainingCostsString = $this->generateSentenceFromArray($trainingCostsStringParts);
+    
+        $message = sprintf(
+            'Training of %s begun at a cost of %s',
+            $unitsToTrainString,
+            $trainingCostsString
+        );
+        
+        return $message;
+    }
+    /**
+     * Generates a string with conjuction from an array of strings
+     *
+     * @param array $stringParts
+     * @return string
+     */
+    private function generateSentenceFromArray($stringParts) {
```

Comment:
This functionality gets used in other places (notably other ActionServices to build the message).

Note to self to:
1) Double-check if there isn't a standard Laravel function for it already, and if not:
2) Move it to a global helper function in like `src/helpers.php` and autoload this file with Composer.
3) Use this helper function wherever it's needed in other ActionServices.

Feel free to pick this up if you want. If not then I can do it myself.

edit: Just checked and Laravel has a `str_replace_last()` function, which could probably be used like:

```PHP
function str_normalize_delimiters(string $string, string $delimiter = ', ', string $lastDelimiter = ' and '): string
{
    return str_replace_last($delimiter, $lastDelimiter, $string);
}
```

Haven't tested it, but my guess is something like this'll work.

edit2: Or honestly just use `str_replace_last()` in the ActionServices, don't really need a wrapper function that does nothing else.